File: Searching/Sahil/JAVA/TernarySearch.java
package searching;

public class TernarySearch {

    public static int ternarySearch(int[] arr, int numberToSearch) {

        int position = -1 ;
        int low = 0, high = arr.length - 1 ;
        while (low <= high) {

            int midFirst = low + (high - low) / 3 ;
            int midSecond = midFirst + (high - low) / 3 ;

            if(arr[midFirst] == numberToSearch) {

                position = midFirst ;
                break ;
            } else if(arr[midSecond] == numberToSearch) {

                position = midSecond ;
                break ;
            } else if (arr[midFirst] > numberToSearch) {
                high = midFirst - 1 ;
            } else if(arr[midSecond] < numberToSearch) {
                low = midSecond + 1 ;
            } else {
                low = midFirst + 1 ;
                high = midSecond - 1 ;
            }

        }
        return position ;
    }

    public static void main(String[] args) {

        int[] arr = {10, 20, 30, 40, 50, 60, 70, 80, 90, 100, 110, 120, 130, 140, 150, 160, 170, 180, 190, 200 } ;
        int numberToSearch = 200 ;
        int searchResult = ternarySearch(arr, numberToSearch) ;

        if(searchResult >= 0) {
            System.out.println("Number found at: " + (searchResult + 1));
        } else {
            System.out.println("Number not found");
        }
    }
}

File: Searching/Sahil/C++/ternarySearch.cpp
/*
Works only with sorted arrays
*/

#include<iostream>
#include<array>
#include<bits/stdc++.h>
using namespace std;

int jumpSearch(int arr[], int size, int number_to_search) {

    int min = 0 ;
    int max = size - 1 ;

    while (min <= max) {

        int midFirst = min + (max - min) /  3 ;
        int midSecond = midFirst + (max - min) / 3 ;

        if(arr[midFirst] == number_to_search) {

            return midFirst ;
        } else if(arr[midSecond] == number_to_search) {

            return midSecond ;
        } else if(arr[midFirst] > number_to_search) {

            max = midFirst - 1 ;
        } else if(arr[midSecond] < number_to_search) {
            
            min = midSecond + 1 ;
        } else {

            min = midFirst + 1 ;
            max = midSecond - 1 ;
        }
    } 

    return -1 ;
}

int main() {

    array<int, 16> array = { 1, 2 , 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16} ;

    int number_to_search = 2 ;

    int search_result = jumpSearch(array.data(), array.size(), number_to_search) ;

    if (search_result >= 0) {

        cout << "got it at: " << search_result + 1 << endl ;
    } else {

        cout << "No such value present" << endl ;
    }
    return 0 ;
}